src/Bicep.LangServer/Completions/BicepCompletionContext.cs

```diff
@@ -632,6 +643,40 @@ private static bool PropertyTypeShouldFlowThrough(List<SyntaxBase> matchingNodes
             return false;
         }
 
+        private static bool ArrayItemTypeShouldFlowThrough(List<SyntaxBase> matchingNodes, (ArrayItemSyntax? node, int index) arrayItemInfo)
+        {
+            if (arrayItemInfo.node is null)
+            {
+                return false;
+            }
+
+            // Array item types should flow through parenthesized and ternary expressions. For examples:
+            // [
+            //   ( | )
+            // ]
+            // [
+            //   conditionA ? | : false
+            // ]
+            // [
+            //   conditionA ? (conditionB ? true : |) : false
+            // ]
+            if ((SyntaxMatcher.IsTailMatch<TernaryOperationSyntax>(matchingNodes) ||
+                 SyntaxMatcher.IsTailMatch<TernaryOperationSyntax, Token>(matchingNodes) ||
+                 SyntaxMatcher.IsTailMatch<ParenthesizedExpressionSyntax>(matchingNodes) ||
+                 SyntaxMatcher.IsTailMatch<ParenthesizedExpressionSyntax, Token>(matchingNodes) ||
+                 SyntaxMatcher.IsTailMatch<ParenthesizedExpressionSyntax, SkippedTriviaSyntax>(
```

Comment:
Could you instead take `matchingNodes` and just return true if the last node that is not a `TernaryOperationSyntax` or a `ParenthesizedExpressionSyntax` is an `ArrayItemSyntax`? I wonder if the current approach would incorrectly match something that ends in a ternary and is inside an array but is not an array item, e.g.,

```bicep
[
  {
    key: ( | )
  }
]
```

Reply:
It will tail match it but the last condition will exclude this case because it checks if all the nodes between the array item type and all but last token are ternary/parentheses nodes.
![Screenshot 2023-03-21 112600](https://user-images.githubusercontent.com/5498623/226656511-4ab0ee05-8699-44b6-8601-0941c0121000.png)

I can add tests for this method so the cases are more clear. I can't remember now what the skipped trivia syntax case was.

Reply:
Ah, I missed that last condition. Could the tail matchers could be skipped in favor of a version of the last condition, for example:
```csharp
matchingNodes.Skip(arrayItemInfo.index + 1).SkipLast(1).All(node => node is TernaryOperationSyntax or ParenthesizedExpressionSyntax) && (matchingNodes.Last() is Token || matchingNodes.Last() is SkippedTriviaSyntax)
```

When I left the comment above, I was thinking of something more iterative, like:
```csharp
for (int i = matchingNodes.Count - 1; i >= 0; i--)
{
    switch (matchingNodes[i])
    {
        case Token when i == matchingNodes.Count - 1:
        case SkippedTriviaSyntax when i == matchingNodes.Count - 1:
        case ParenthesizedExpressionSyntax:
        case TernaryOperationSyntax:
            continue;
        case ArrayItemSyntax:
            return true;
        default:
            return false;
    }
}
```

But the concern I raised is already covered, so you should go with whatever approach strikes you as most readable.